executor/aggregate.go

```diff
@@ -45,39 +96,482 @@ type HashAggExec struct {
+			w.outputCh <- &AfFinalResult{chk: result, giveBackCh: w.finalResultHolderCh}
+			result, ok = <-w.finalResultHolderCh
+			if !ok {
+				return
+			}
+			result.Reset()
+		}
+	}
+}
+
+func (w *HashAggFinalWorker) run(ctx sessionctx.Context, waitGroup *sync.WaitGroup) {
+	defer func() {
+		waitGroup.Done()
+	}()
+	sc := ctx.GetSessionVars().StmtCtx
+	if err := w.fetchIntermData(sc); err != nil {
+		w.outputCh <- &AfFinalResult{err: errors.Trace(err)}
+	}
+	w.getFinalResult(sc)
+}
+
 // Next implements the Executor Next interface.
 func (e *HashAggExec) Next(ctx context.Context, chk *chunk.Chunk) error {
+	chk.Reset()
+	if e.doesUnparallelExec {
+		return errors.Trace(e.unparallelExec(ctx, chk))
+	}
+	return errors.Trace(e.parallelExec(ctx, chk))
+}
+
+func (e *HashAggExec) fetchChildData(ctx context.Context) {
+	var (
+		input *HashAggInput
+		chk   *chunk.Chunk
+		ok    bool
+		err   error
+	)
+	defer func() {
+		for _, w := range e.partialWorkers {
+			close(w.inputCh)
+		}
+	}()
+	for {
+		select {
+		case <-e.finishCh:
+			return
+		case input, ok = <-e.inputCh:
+			if !ok {
+				return
+			}
+			chk = input.chk
+		}
+		err = e.children[0].Next(ctx, chk)
+		if err != nil {
+			e.finalOutputCh <- &AfFinalResult{err: errors.Trace(err)}
+			return
+		}
+		if chk.NumRows() == 0 {
+			return
+		}
+		input.giveBackCh <- chk
+	}
+}
+
+func (e *HashAggExec) waitPartialWorkerAndCloseOutputChs(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	for _, ch := range e.partialOutputChs {
+		close(ch)
+	}
+}
+
+func (e *HashAggExec) waitFinalWorkerAndCloseFinalOutput(waitGroup *sync.WaitGroup) {
+	waitGroup.Wait()
+	close(e.finalOutputCh)
+}
+
+// parallelExec executes hash aggregate algorithm parallelly.
+func (e *HashAggExec) parallelExec(ctx context.Context, chk *chunk.Chunk) error {
+	if !e.prepared {
+		go e.fetchChildData(ctx)
+
+		partialWorkerWaitGroup := &sync.WaitGroup{}
+		partialWorkerWaitGroup.Add(len(e.partialWorkers))
+		for i := range e.partialWorkers {
+			go e.partialWorkers[i].run(e.ctx, partialWorkerWaitGroup, len(e.finalWorkers))
+		}
+		go e.waitPartialWorkerAndCloseOutputChs(partialWorkerWaitGroup)
+
+		finalWorkerWaitGroup := &sync.WaitGroup{}
+		finalWorkerWaitGroup.Add(len(e.finalWorkers))
+		for i := range e.finalWorkers {
+			go e.finalWorkers[i].run(e.ctx, finalWorkerWaitGroup)
+		}
+		go e.waitFinalWorkerAndCloseFinalOutput(finalWorkerWaitGroup)
+
+		e.prepared = true
+	}
+	for {
+		result, ok := <-e.finalOutputCh
+		if !ok || result.err != nil || result.chk.NumRows() == 0 {
+			if result != nil {
+				return errors.Trace(result.err)
+			}
+			if e.isInputNull && e.defaultVal != nil {
+				chk.Append(e.defaultVal, 0, 1)
+			}
+			e.isInputNull = false
+			return nil
+		}
+		e.isInputNull = false
+		chk.SwapColumns(result.chk)
+		// Put result.chk back to the corresponded final worker's finalResultHolderCh.
+		result.giveBackCh <- result.chk
+		// todo: store the result that chk.numrows() < e.maxChunkSize
```

Comment:
Maybe this todo can be removed. The situation mentioned here only exists when the chunk is the last chunk returned by the final workers, and since we do not have many final workers, it is not worth to make every chunk be limited with num rows equals to max chunk size.